### PR TITLE
Rewrite rpc/client and rpc/send

### PR DESCRIPTION
**Pull request by tamird**
Addresses #1521.

Somewhat orthogonally, this adds a 3-second TLS timeout.

### CODE REVIEW

File: rpc/client.go [hunk 1]
```diff
@@ -106,15 +107,35 @@ func NewClient(addr net.Addr, opts *retry.Options, context *Context) *Client {
...
+		retryOpts = clientRetryOptions
+	} else {
+		retryOpts = *opts
+	}
+	retryOpts.Tag = fmt.Sprintf("client %s connection", c.addr)
+	retryOpts.Stopper = context.Stopper
+
+	reconnect := func() error {
+		return retry.WithBackoff(retryOpts, func() (retry.Status, error) {
+			return retry.Continue, c.connect(tlsConfig)
```
Comment (petermattis) [inline rpc/client.go]: Won't this retry indefinitely?

File: rpc/client.go [hunk 2]
```diff
@@ -234,19 +224,44 @@ func (c *Client) Close() {
...
+				c.offset.Reset()
+				c.mu.Unlock()
+
+				failures++
+
+				log.Infof("client %s heartbeat failed %d times : %v; marked unhealthy", c.Addr(), failures, err)
+
+				if failures > 3 {
+					log.Infof("client %s reconnecting after %d heartbeat failures...", c.Addr(), failures)
+					if err := reconnect(); err != nil {
```
Comment (petermattis) [inline rpc/client.go]: So once an RPC client is created we'll keep trying to open connections to the remote node until the local node is restarted? This package is woefully under-documented. I think it would be good to call out some of the behavioral expectations in a comment at the top of the file.

@spencerkimball also needs to take a look at this. I'm familiar with the `rpc/codec` code, but this code is originally his.

File: rpc/client.go (hunk 1 above)
Comment (tamird) [inline rpc/client.go]: Yeah. For the most part, I think we want this (until we have rebalancing, at least), since the thing on the other end is known to have been alive at some point.

File: rpc/tls.go [hunk 3]
```diff
@@ -44,6 +45,10 @@ func tlsListen(network, address string, config *tls.Config) (net.Listener, error
 	return tls.Listen(network, address, config)
 }
 
+var defaultDialer = net.Dialer{
+	Timeout: 3 * time.Second,
```
Comment (petermattis) [inline rpc/tls.go]: What's the timeout with the standard dialer?

Comment (tamird) [inline rpc/tls.go]: Indefinite.

File: rpc/client.go (hunk 1 above)
Comment (petermattis) [inline rpc/client.go]: My point was more direct: by always returning `retry.Continue` you'll never exit the retry loop. Am I missing something?

Comment (tamird) [inline rpc/client.go]: oh jeez, you're right. I'll fix this.

File: rpc/client.go [hunk 4]
```diff
@@ -58,7 +59,7 @@ var clientRetryOptions = retry.Options{
 
 // init creates a new client RPC cache.
 func init() {
-	clients = map[string]*Client{}
+	clients = make(map[string]*Client)
```
Comment (petermattis) [inline rpc/client.go]: There is no real reason to prefer `make` over the empty map construction. We use the latter frequently throughout the code. Let's leave this alone.

File: rpc/client.go (hunk 1 above)
Comment (tamird) [inline rpc/client.go]: fixed

File: rpc/client.go (hunk 4 above)
Comment (tamird) [inline rpc/client.go]: removed

File: rpc/client.go [hunk 5]
```diff
@@ -106,15 +107,38 @@ func NewClient(addr net.Addr, opts *retry.Options, context *Context) *Client {
 		Closed:       make(chan struct{}),
 		clock:        context.localClock,
 		remoteClocks: context.RemoteClocks,
-		cached:       !context.DisableCache,
 	}
-	if !context.DisableCache {
-		clients[c.Addr().String()] = c
+	clients[c.Addr().String()] = c
+
+	tlsConfig, tlsErr := context.GetClientTLSConfig()
```
Comment (MBerhault) [inline rpc/client.go]: you might as well error out right away (well, log an error), if tlsErr is not nil, we shouldn't be retrying.

File: rpc/client.go [hunk 6]
```diff
@@ -94,9 +94,10 @@ type Client struct {
 // method is invoked.
 func NewClient(addr net.Addr, opts *retry.Options, context *Context) *Client {
```
Comment (MBerhault) [inline rpc/client.go]: this has entirely too much logic in it to not return an error.
We should either have NewClient return errors properly, or have just the basic initialization. Connection would be done later.

File: rpc/client.go (hunk 5 above)
Comment (MBerhault) [inline rpc/client.go]: I would actually pass the context to the dialer, and have those call GetClientTLSConfig. Also, calling it more than once doesn't matter, it does lazy loading of the actual configs, so everything after the first successful call is cheap.

File: rpc/client.go [hunk 7]
```diff
@@ -122,7 +146,7 @@ func NewClient(addr net.Addr, opts *retry.Options, context *Context) *Client {
 			close(c.Ready)
```
Comment (tschottdorf) [inline rpc/client.go]: the `Ready` channel loses its functionality with this change because it'll only work in the intended way until the client first encounters connection issues. You'll need to replace it when the peer goes dead and close it on each successful attempt. Plus access to it must be behind the lock, so it needs to become private.

File: server/raft_transport.go [hunk 8]
```diff
@@ -190,8 +190,10 @@ func (t *rpcTransport) processQueue(raftNodeID proto.RaftNodeID) {
 		}
 
 		if !client.IsHealthy() {
-			log.Warningf("raft client for node %d unhealthy", nodeID)
-			return
+			if log.V(1) {
```
Comment (tschottdorf) [inline server/raft_transport.go]: this will be spammy, the old code would return which would let the next message wait for the client to become ready (which has its own issues right now, see above)

File: rpc/client.go (hunk 7 above)
Comment (tschottdorf) [inline rpc/client.go]: A test for this would be good as well.

File: server/raft_transport.go (hunk 8 above)
Comment (tschottdorf) [inline server/raft_transport.go]: (should add that the old code would drop the outgoing message queue, but if you miss a couple of Raft messages, then that's likely not a bad idea anyway).

**Comment by tschottdorf** [comment]
I added some comments to #1521 which are relevant here.

File: server/raft_transport.go (hunk 8 above)
Comment (bdarnell) [inline server/raft_transport.go]: In general we do want to drop raft messages if a peer is known to be unhealthy. There's no point in unleashing a ton of queued heartbeats as soon as a node comes back up.

File: rpc/client.go [hunk 9]
```diff
@@ -106,15 +107,38 @@ func NewClient(addr net.Addr, opts *retry.Options, context *Context) *Client {
...
 		remoteClocks: context.RemoteClocks,
-		cached:       !context.DisableCache,
 	}
-	if !context.DisableCache {
-		clients[c.Addr().String()] = c
+	clients[c.Addr().String()] = c
+
+	tlsConfig, tlsErr := context.GetClientTLSConfig()
+
+	connect := func() error {
```
Comment (bdarnell) [inline rpc/client.go]: This function appears to only be called once, so it could just be inlined into the RunWorker closure. But I agree with Marc that we should probably either add an error return to NewClient and fail immediately if we can't load the TLS configs, or delay the call to GetClientTLSConfig until we're actually trying to connect. I prefer the latter since everything's cached; we generally don't have great error-handling options at the point where we're calling NewClient.

**Comment by bdarnell** [comment]
This looks like a step in the right direction to me, although we do need to address @tschottdorf's comment on the `Ready` channel.

**Comment by tamird** [comment]
I've pushed another two commits which amount to a rewrite of rpc/client.go and rpc/send.go respectively; I believe this addresses the shortcomings identified here and in #1521. PTAL

File: rpc/client.go [hunk 10]
```diff
@@ -90,219 +95,234 @@ type Client struct {
...
-		c.mu.Unlock()
-		log.Warningf("client %s unhealthy after %s", c.Addr(), heartbeatInterval)
-		return util.Errorf("client timeout")
 	case <-c.Closed:
 		return errClosed
+	case <-call.Done:
+		if err := call.Error; err != nil {
+			return err
+		}
+	case <-time.After(2 * heartbeatInterval):
```
Comment (tschottdorf) [inline rpc/client.go]: when the client is unhealthy, shouldn't we retry a little more frequently? After all, we're only expecting one client per peer.

File: rpc/send.go [hunk 11]
```diff
@@ -153,121 +142,72 @@ func Send(opts Options, method string, addrs []net.Addr, getArgs func(addr net.A
...
-		// Start clients up to N.
-		for ; index < N; index++ {
-			args := getArgs(clients[index].Addr())
-			if args == nil {
-				helperChan <- util.Errorf("nil arguments returned for client %s", clients[index].Addr())
-				continue
-			}
-			reply := getReply()
+	// Start clients up to N.
+	for _, addr := range orderedAddrs { // TRIM: maybe do fewer [:opts.N]
```
Comment (tschottdorf) [inline rpc/send.go]: why not trim this now? If this means you're always sending to all replicas that's not something we'd want to check in.

**Comment by tschottdorf** [comment]
what happened with the idea of removing `NewClient` generally? Out of scope or did you find it was problematic?

**Comment by tamird** [comment]
It was basically a choice between having `NewClient` return an error and implementing `rpc.Go`, and I wanted to make sure we surface connection errors synchronously.

Taking this further, it's probably not necessary to have any of the signaling we have on the client now; once `connect()` succeeds for the first time, we already know the other end is at least a Go program speaking go/rpc (but not necessarily cockroach or even cockroach's protobuf-based codec). It's probably safe to just start sending requests at the moment of successful return from `NewClient`.

File: rpc/send.go (hunk 11 above)
Comment (tamird) [inline rpc/send.go]: implemented

File: rpc/client.go (hunk 10 above)
Comment (tamird) [inline rpc/client.go]: yeah, we probably should; I'll do that in a subsequent PR since it's orthogonal to this change in a sense

File: rpc/client.go [hunk 12]
```diff
@@ -66,19 +68,21 @@ func init() {
...
-	mu           sync.Mutex // Mutex protects the fields below
-	*rpc.Client             // Embedded RPC client
-	addr         net.Addr   // Remote address of client
-	lAddr        net.Addr   // Local address of client
-	healthy      bool
-	offset       proto.RemoteOffset // Latest measured clock offset from the server
+	rpcClient *rpc.Client
+
+	Closed        chan struct{}
+	healthy       atomic.Value
```
Comment (bdarnell) [inline rpc/client.go]: Comment that this holds a `chan struct{}`.

File: rpc/client.go [hunk 13]
```diff
@@ -91,219 +95,230 @@ type Client struct {
...
-func NewClient(addr net.Addr, opts *retry.Options, context *Context) *Client {
+// closed if the client's Close() method is invoked.
+func NewClient(addr net.Addr, context *Context) (*Client, error) {
 	clientMu.Lock()
+	defer clientMu.Unlock()
+
+	network := addr.Network()
+	address := addr.String()
+
+	key := network + address
```
Comment (bdarnell) [inline rpc/client.go]: Mainly a theoretical concern, but this needs a delimiter to be unambiguous since some networks are prefixes of each other ("tcp" vs "tcp4" and "tcp6").

File: rpc/client.go [hunk 14]
```diff
@@ -91,219 +95,230 @@ type Client struct {
...
+		tlsConfig:     tlsConfig,
+		clock:         context.localClock,
+		remoteClocks:  context.RemoteClocks,
 	}
-	if !context.DisableCache {
-		clients[c.Addr().String()] = c
+
+	c.healthy.Store(make(chan struct{}))
+
+	if err = c.connect(); err != nil {
```
Comment (bdarnell) [inline rpc/client.go]: We should release `clientMu` before calling `connect()`.

File: rpc/client.go [hunk 15]
```diff
@@ -91,219 +95,230 @@ type Client struct {
...
+		return nil, err
 	}
-	clientMu.Unlock()
+
+	clients[key] = c
```
Comment (bdarnell) [inline rpc/client.go]: When `DisableCache` is false we should not store `c` in `clients`.

File: rpc/client.go [hunk 16]
```diff
@@ -91,219 +95,230 @@ type Client struct {
...
+		<-context.Stopper.ShouldStop()
 
-			// Launch periodic heartbeat. This blocks until the client is to be closed.
-			c.runHeartbeat(context.Stopper)
-		}
 		c.Close()
 	})
-	return c
+
+	go func() {
```
Comment (bdarnell) [inline rpc/client.go]: This looks like it should be a `Stopper.RunWorker` instead of a bare `go func()`

Comment (tamird) [inline rpc/client.go]: It's intentionally not a `RunWorker` because I wanted to avoid selecting on `c.Closed` **and** `stopper.ShouldStop()` everywhere; instead all the internals select on `c.Closed` and the `RunWorker` above waits on `ShouldStop()` and then calls `c.Close()`.

File: rpc/client.go (hunk 15 above)
Comment (tamird) [inline rpc/client.go]: Why not? The usage pattern suggests that callers only care that they get a fresh client, not that nobody else may reuse this client.

File: rpc/client.go (hunk 14 above)
Comment (tamird) [inline rpc/client.go]: You're right, fixing.

File: rpc/client.go (hunk 13 above)
Comment (tamird) [inline rpc/client.go]: fixing

File: rpc/client.go (hunk 12 above)
Comment (tamird) [inline rpc/client.go]: fixed

**Comment by tschottdorf** [comment]
unless I'm overlooking something, a client may never give up even if it's not used. Can you add a TODO for that?

File: rpc/client.go [hunk 17]
```diff
@@ -91,219 +95,230 @@ type Client struct {
...
-// connect dials the connection in a backoff/retry loop.
-func (c *Client) connect(opts *retry.Options, context *Context) error {
-	// Attempt to dial connection.
-	retryOpts := clientRetryOptions
-	if opts != nil {
-		retryOpts = *opts
+// Go wraps net/rpc.Client.Go and provides additional cockroach-specific error
+// handling.
+func (c *Client) Go(serviceMethod string, args interface{}, reply interface{}, done chan *rpc.Call) *rpc.Call {
+	call := c.rpcClient.Go(serviceMethod, args, reply, nil)
```
Comment (bdarnell) [inline rpc/client.go]: There is a race between reading `c.rpcClient` here (which could be in any thread) and writing to it in `connect` from the heartbeat thread.

File: rpc/client.go [hunk 18]
```diff
@@ -91,219 +95,230 @@ type Client struct {
...
-			return err
-		}
+	return newCall
+}
 
-		return nil
+func (c *Client) connect() error {
+	conn, err := tlsDialHTTP(c.network, c.remoteAddress, c.tlsConfig)
+	if err != nil {
+		return err
```
Comment (bdarnell) [inline rpc/client.go]: We should clear `c.rpcClient` if we fail to (re-)connect, and `Go` should be updated to handle a nil rpcClient gracefully.

File: rpc/client.go [hunk 19]
```diff
@@ -91,219 +95,230 @@ type Client struct {
...
-		log.Infof("client %s starting heartbeat", c.Addr())
-	}
-
+func (c *Client) runHeartbeat(retryOpts retry.Options) {
+	var err error
 	for {
+		for r := retry.Start(retryOpts); r.Next(); {
+			// reconnect unless the error is nil or definitely temporary
+			if netErr, ok := err.(net.Error); err != nil && !(ok && netErr.Temporary()) {
+				_ = c.connect()
```
Comment (bdarnell) [inline rpc/client.go]: If connect() returns an error, we should log it, setUnhealthy, and continue. The error from connect() is probably going to be more useful to log than the error from the heartbeat.

File: rpc/client.go [hunk 20]
```diff
@@ -91,219 +95,230 @@ type Client struct {
...
-	if log.V(2) {
-		log.Infof("client %s starting heartbeat", c.Addr())
-	}
-
+func (c *Client) runHeartbeat(retryOpts retry.Options) {
+	var err error
 	for {
+		for r := retry.Start(retryOpts); r.Next(); {
+			// reconnect unless the error is nil or definitely temporary
+			if netErr, ok := err.(net.Error); err != nil && !(ok && netErr.Temporary()) {
```
Comment (bdarnell) [inline rpc/client.go]: I would be surprised if we got a temporary error here, since this error is for the heartbeat rpc rather than connect. I don't think you can get a temporary error once a connection is established unless you call `SetDeadline`.

File: rpc/client.go [hunk 21]
```diff
@@ -91,219 +95,230 @@ type Client struct {
...
 }
 
+// LocalAddr returns the local address of the client.
+func (c *Client) LocalAddr() net.Addr {
+	return util.MakeUnresolvedAddr(c.network, c.localAddress)
+}
+
+// RemoteAddr returns remote address of the client.
+func (c *Client) RemoteAddr() net.Addr {
+	return util.MakeUnresolvedAddr(c.network, c.remoteAddress)
```
Comment (bdarnell) [inline rpc/client.go]: Our inputs for both the local and remote addresses are net.Addr objects; it seems a little silly to split those up into network/address pairs and then reconstruct them into "unresolved" addresses. `TCPAddr` has more than just two fields (for ipv6) so converting to/from an unresolved address involves constructing a string and parsing it.

File: rpc/send.go [hunk 22]
```diff
@@ -154,121 +143,92 @@ func Send(opts Options, method string, addrs []net.Addr, getArgs func(addr net.A
...
-		// Start clients up to N.
-		for ; index < N; index++ {
-			args := getArgs(clients[index].Addr())
-			if args == nil {
-				helperChan <- util.Errorf("nil arguments returned for client %s", clients[index].Addr())
-				continue
-			}
-			reply := getReply()
+	sendFn := func(addr net.Addr) {
+		// NewClient might block while connecting, but connecting in parallel
```
Comment (bdarnell) [inline rpc/send.go]: I don't understand this comment.

File: rpc/send.go [hunk 23]
```diff
@@ -125,28 +124,18 @@ func Send(opts Options, method string, addrs []net.Addr, getArgs func(addr net.A
...
+
+	var orderedAddrs []net.Addr
 	switch opts.Ordering {
 	case OrderStable:
-		for _, addr := range addrs {
-			clients = append(clients, NewClient(addr, nil, context))
-		}
+		orderedAddrs = addrs
 	case OrderRandom:
-		// Randomly permute order, but keep known-unhealthy clients last.
```
Comment (bdarnell) [inline rpc/send.go]: Why did you remove the "healthy first" optimization? I'm not seeing anything that does an equivalent job.

File: rpc/send.go [hunk 24]
```diff
@@ -154,121 +143,92 @@ func Send(opts Options, method string, addrs []net.Addr, getArgs func(addr net.A
...
-				}
-				replies = append(replies, t)
-				if successes == opts.N {
-					return replies, nil
+					log.Infof("%s: successful reply: %+v", method, call.Reply)
 				}
+
+				replies = append(replies, call.Reply)
 			}
 		case <-time.After(opts.SendNextTimeout):
```
Comment (bdarnell) [inline rpc/send.go]: Preexisting issue: the time.After should be a variable that lives outside the loop so we don't reset the clock each time a response comes in.

File: rpc/send.go [hunk 25]
```diff
@@ -154,121 +143,92 @@ func Send(opts Options, method string, addrs []net.Addr, getArgs func(addr net.A
...
-			go sendOneFn(clients[index], opts.Timeout, method, args, reply, helperChan)
+			trace.Event(fmt.Sprintf("sending to %s", addr))
+			go sendOneFn(client, opts.Timeout, method, args, getReply(), done)
+		} else {
+			done <- &rpc.Call{Error: newRPCError(util.Errorf("nil arguments returned for client %s", addr))}
 		}
-		// Wait for completions.
+	}
+
+	// Start clients up to opts.N.
```
Comment (bdarnell) [inline rpc/send.go]: It looks like opts.N is always 1, and i'm not sure when we'd ever change that. Should we just rip out all the support for RPCs that expect multiple replies?

File: rpc/client.go [hunk 26]
```diff
@@ -91,219 +95,230 @@ type Client struct {
...
+		newCall.Reply = finishedCall.Reply
+		newCall.Error = finishedCall.Error
 
-		conn, err := tlsDialHTTP(c.addr.Network(), c.addr.String(), tlsConfig)
-		if err != nil {
-			// Retry if the error is temporary, otherwise fail fast.
-			if t, ok := err.(net.Error); ok && t.Temporary() {
-				if log.V(1) {
-					log.Warning(err)
+		if newCall.Error != nil {
```
Comment (bdarnell) [inline rpc/client.go]: This feels like the wrong place for this error wrapping. I'd rather have the code in send.go that checks for Retryable be aware of io.ErrUnexpectedEOF as a special case.

File: rpc/client.go [hunk 27]
```diff
@@ -91,219 +95,230 @@ type Client struct {
...
+			// From net/rpc/client.go; remote hung up. Our heartbeat runloop will
+			// reconnect, so this retryable.
+			if newCall.Error == io.ErrUnexpectedEOF {
+				newCall.Error = newRPCError(newCall.Error)
+			}
+		} else {
+			// Verify response data integrity if this is a proto response.
```
Comment (bdarnell) [inline rpc/client.go]: This seems like it should be the job of the DistSender instead of the rpc transport. This code is shared by both DistSender and the raft transport so it should only do things that are applicable to both.

File: rpc/client.go (hunk 15 above)
Comment (bdarnell) [inline rpc/client.go]: The one place that sets `DisableCache` also sets `Context.RemoteClock` differently from other callers, so it would be inappropriate to reuse the client.

In fact, the entire Context object should really be considered as part of the key (or the cache should be stored in the Context itself instead of in globals); it's not safe to create a client with one Context and then reuse it with another since the Context is only considered the first time we see a given address.

File: rpc/client.go (hunk 16 above)
Comment (bdarnell) [inline rpc/client.go]: I don't think you'd have to select on both channels everywhere if this was a `RunWorker`. The previous worker will ensure that Closed is always closed when ShouldStop is. The reason to make it a worker is that then the Stopper will ensure that this goroutine has exited before calling the Closers (possibly leading to panics if runHeartbeat is too slow to respond to the signal of c.Closed).

File: rpc/client.go [hunk 28]
```diff
@@ -91,219 +100,230 @@ type Client struct {
...
+	close(c.healthy.Load().(chan struct{}))
 }
 
-// RemoteOffset returns the most recently measured offset of the client clock
-// from the remote server clock.
-func (c *Client) RemoteOffset() proto.RemoteOffset {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	return c.offset
+func (c *Client) setUnhealthy() {
```
Comment (tschottdorf) [inline rpc/client.go]: `set{Unhealthy,Healthy}` can only ever be called from one place because they don't lock (and that place is the heartbeat loop) so maybe they should live in there for the time being.

File: rpc/send.go (hunk 23 above)
Comment (tschottdorf) [inline rpc/send.go]: `IsHealthy` went away. Is the following code cheap? If so, might be a good way to re-implement it. I'm also wondering whether we really need `c.isHealthy` (a bool) since we only sequentially touch the channel from a single function which can easily remember whether the channel was closed or not.
```
select {
  <-c.Healthy():
     return true
}
return false
```

File: rpc/send.go (hunk 25 above)
Comment (tschottdorf) [inline rpc/send.go]: I think that's a good idea.

**Comment by tamird** [comment]
@tschottdorf you're correct, added a TODO

File: rpc/send.go (hunk 22 above)
Comment (tamird) [inline rpc/send.go]: Yeah, this comment's placement is weird. It made sense in a previous iteration. Fixing.

File: rpc/client.go (hunk 16 above)
Comment (tamird) [inline rpc/client.go]: oh, I see. I'll make them both workers

File: rpc/client.go (hunk 17 above)
Comment (tamird) [inline rpc/client.go]: fixed

File: rpc/client.go (hunk 19 above)
Comment (tamird) [inline rpc/client.go]: done

File: rpc/client.go (hunk 18 above)
Comment (tamird) [inline rpc/client.go]: I think the built-in stuff already takes care of this for us. If the connection is dead, `codec.WriteRequest` will fail and a call object with a populated error will be returned.

File: rpc/client.go (hunk 20 above)
Comment (tamird) [inline rpc/client.go]: Yeah, I'm not sure when it would happen, but I don't think it's harmful to have this here.

File: rpc/send.go (hunk 24 above)
Comment (tamird) [inline rpc/send.go]: Yeah, this is a weird one. It wasn't clear to me which way the intended semantics were. Gonna leave it for now.

File: rpc/client.go (hunk 26 above)
Comment (tamird) [inline rpc/client.go]: How come? `rpc.NewClient` is a public method, which means it's used by more than just send.go, and the semantics of the one-client-that-keeps-reconnecting are what makes this error retryable. Why not provide that knowledge it to all clients?

File: rpc/client.go (hunk 28 above)
Comment (tamird) [inline rpc/client.go]: `set{Unh,H}healthy` are both atomic, though.

File: rpc/send.go (hunk 25 above)
Comment (tamird) [inline rpc/send.go]: Sure, I'll do that in a bit.

Comment (tamird) [inline rpc/send.go]: actually, that doesn't buy us much. I'm going to leave it in.

File: rpc/client.go (hunk 27 above)
Comment (tamird) [inline rpc/client.go]: What about gossip? It also shares this code.

File: rpc/client.go (hunk 28 above)
Comment (tschottdorf) [inline rpc/client.go]: but only because the caller holds the lock. If you add comments that they presuppose the lock being held, it's ok for me.

Comment (tamird) [inline rpc/client.go]: What? The primitive being used for `c.healthy` is an `atomic.Value` - the methods are safe no matter what

Comment (tamird) [inline rpc/client.go]: Oh, you mean because of `isHealthy`. Sorry, you're right. I'll get that part fixed (hopefully by removing isHealthy)

File: rpc/send.go (hunk 24 above)
Comment (tschottdorf) [inline rpc/send.go]: it makes more sense outside of the loop since all requests are sent in parallel.

**Comment by tamird** [comment]
I'm going to hold off on further work here for now. @bdarnell's comment about the context being part of the cache key got me thinking about this more broadly; it seems really awkward to have the base rpc client do these heartbeat+clock updates when (I think) only the dist sender cares about them (raft and gossip do not, AFAICT).

I'll file an issue to track that refactor, but probably won't tackle it for a little while.

**Comment by tschottdorf** [comment]
which comment are you referring to?

Clock updates are relevant for all connections which talk to other nodes. Heartbeats are just keep-alives (which is usually also relevant for all connections in a cluster). And your change definitely looked like an improvement. Why abandon it completely?

File: rpc/client.go (hunk 27 above)
Comment (bdarnell) [inline rpc/client.go]: Gossip doesn't care about this Verify operation either - it's unique to our KV `Request/ResponseHeader`.

File: rpc/send.go (hunk 23 above)
Comment (bdarnell) [inline rpc/send.go]: Yeah, I think a non-blocking select on `c.Healthy()` would be cheap enough.

File: rpc/client.go (hunk 26 above)
Comment (bdarnell) [inline rpc/client.go]: Fair enough; I was getting mixed up about the layering here. But wrapping the channel in an extra goroutine seems like an awfully heavyweight way to tell the client that an error is retryable. The problem is that type-asserting to `util.Retryable` is inflexible and can only work with error types we define ourselves. Instead, we should follow the example of `os.IsNotExist()` and friends, and provide a `CanRetry(error)` function which can understand the various ways in which an error can be retryable (`util.Retryable`, `net.Error`, etc).

Comment (tamird) [inline rpc/client.go]: Sounds good to me. One problem with this is that the "client connection died" error is retryable _only_ if it comes from this special sauce client that does automatic reconnection, so I don't think it'll help us avoid the extra channel and goroutine in this case.
However, I can probably implement this using `client.Call` rather than `client.Go` which will save one goroutine. I'll do that now and reopen this.

File: rpc/client.go (hunk 21 above)
Comment (tamird) [inline rpc/client.go]: fixed this by retaining the `conn` object; these methods just delegate to it.

File: rpc/send.go (hunk 23 above)
Comment (tamird) [inline rpc/send.go]: coming up

File: rpc/client.go (hunk 27 above)
Comment (tamird) [inline rpc/client.go]: fixed

File: rpc/client.go (hunk 28 above)
Comment (tamird) [inline rpc/client.go]: fixed

**Comment by tamird** [comment]
OK, I think I've addressed everything. @bdarnell I've had to revert to holding the cache mutex while `connect()`ing to avoid putting a dead client into the cache (which would not have the reconnect loop running).

**Comment by bdarnell** [comment]
It's unacceptable to block `clientMu` while we're doing network I/O. I think we'll have to get rid of the connect() call in NewClient and only connect from the retry loop. If we want to be able to return a connect error from NewClient we'll need to introduce some way to return it from the retry loop to the caller, but I'm not sure that's worth it since we plan to make client creation invisible anyway.

**Comment by tamird** [comment]
> since we plan to make client creation invisible anyway.

I don't think client creation can ever be invisible precisely for the same reason that `connect()` has to block - in the event of configuration errors we want to fail eagerly. That's the major change in this PR: `NewClient` returns an error.

**Comment by bdarnell** [comment]
The way I see it the major change in this PR is to make client creation more invisible, by removing the notion of Ready (a one-time transition that required creation of multiple client objects over time) in favor of Healthy (a reversible transition).

I don't think failing eagerly is desirable in this case. If we connect() from NewClient from the retry loop then in addition to the contention on `clientMu` we've also lost the ability to impose a backoff in the event of a configuration error. As soon as one connection attempt finishes the lock will be released and another attempt will start.

**Comment by tamird** [comment]
Why would we want to have backoff on configuration errors? The caller should either immediately give up or panic (if we statically expect never to see a class of configuration error at a particular call site).

**Comment by bdarnell** [comment]
In the event of a configuration error, we can do one of three things: return an error, insert an unhealthy client in the pool, or panic. If we return an error then the proximate caller may give up but we have no control over other tasks in the same process so something else may hit the same configuration error and try again in an uncontrolled way. Ultimately, all errors can and will be retried at some level (unless we panic), so the rpc/client package should be responsible for managing those retries.

**Comment by tamird** [comment]
> If we return an error then the proximate caller may give up but we have no control over other tasks in the same process so something else may hit the same configuration error and try again in an uncontrolled way.

I don't understand. Are you saying we have no control over someone ignoring a returned error?

> Ultimately, all errors can and will be retried at some level (unless we panic), so the rpc/client package should be responsible for managing those retries.

What happens when a node disappears permanently?

**Comment by bdarnell** [comment]
> I don't understand. Are you saying we have no control over someone ignoring a returned error?

I'm saying there is no singular "someone". In a server, most use of NewClient is in rpc.Send, which does not have any way to remember a past failure from NewClient (nor does it panic).

> What happens when a node disappears permanently?

How do you tell if it's permanent? In general you won't be able to tell from the type of error; it'll just look like a node that's temporarily down. And whatever we do for permanently-gone nodes, we need to be able to do it for clients that are already established, not just those that are yet to start. We need GC of idle rpc.Clients, and then we just wait for the rebalancer to move all traffic from the downed node so no one else tries to connect.

File: gossip/client.go [hunk 29]
```diff
@@ -84,9 +84,6 @@ func (c *client) start(g *Gossip, done chan *client, context *rpc.Context, stopp
 		// Start gossipping and wait for disconnect or error.
 		c.lastFresh = time.Now().UnixNano()
 		c.err = c.gossip(g, stopper)
-		if c.err != nil {
-			c.rpcClient.Close()
```
Comment (bdarnell) [inline gossip/client.go]: I think we still need to close the client if we're bootstrapping (and `context.DisableCache` is true).

File: rpc/client.go [hunk 30]
```diff
@@ -91,219 +97,224 @@ type Client struct {
...
-		}
-		c.Close()
-	})
-	return c
-}
+	c.healthy.Store(make(chan struct{}))
+	// Must store junk in here so `Load()` doesn't blow up
+	c.conn.Store(internalConn{})
+
+	clients[key] = c
```
Comment (bdarnell) [inline rpc/client.go]: We should not cache clients that were created with `context.DisableCache=true` (because in the case of gossip they have an incorrect value of `context.RemoteClocks`), unless we can consider the context as a part of the cache key.

File: rpc/client.go [hunk 31]
```diff
@@ -91,219 +97,224 @@ type Client struct {
...
+	}
+	if done == nil {
+		newCall.Done = make(chan *rpc.Call, 1)
+	} else {
+		newCall.Done = done
 	}
 
-	return util.Errorf("system is stopping")
-}
+	go func() {
```
Comment (bdarnell) [inline rpc/client.go]: This wrapper still feels unnecessary to me; why can't we just update everything that knows to retry on a newRPCError to also retry on ErrUnexpectedEOF?

File: rpc/client.go (hunk 30 above)
Comment (tamird) [inline rpc/client.go]: fixed (locally)

File: rpc/client.go [hunk 32]
```diff
@@ -91,219 +97,224 @@ type Client struct {
...
+	tlsConfig, err := context.GetClientTLSConfig()
+	if err != nil {
+		log.Fatal(err)
+	}
+
 	c := &Client{
-		addr:         addr,
-		Ready:        make(chan struct{}),
 		Closed:       make(chan struct{}),
+		addr:         key,
```
Comment (bdarnell) [inline rpc/client.go]: It seems better to save the input `net.Addr` than the "unresolved" addr we converted it to.

File: rpc/client.go [hunk 33]
```diff
@@ -91,219 +97,224 @@ type Client struct {
...
+	}
+	setUnhealthy := func() {
+		if isHealthy {
+			isHealthy = false
+			c.healthy.Store(make(chan struct{}))
+		}
 	}
 
+	// initial connection
+	connErr := c.connect()
```
Comment (bdarnell) [inline rpc/client.go]: If the initial connection fails, we don't log its error and we retry immediately without waiting for the initial backoff. What if instead we started with `connErr := errors.New("not started yet")`? I think that would do the right thing (i.e. an immediate connection attempt inside the loop, and every attempted connection failure would be logged and retried on schedule).

File: rpc/client.go (hunk 31 above)
Comment (tamird) [inline rpc/client.go]: fixed

File: gossip/client.go (hunk 29 above)
Comment (tamird) [inline gossip/client.go]: fixed

File: rpc/client.go (hunk 33 above)
Comment (tamird) [inline rpc/client.go]: fixed

**Comment by bdarnell** [comment]
LGTM

File: rpc/client.go (hunk 32 above)
Comment (tamird) [inline rpc/client.go]: ah. this was done because the dialer takes strings (rather than a `net.Addr`), and using an UnresolvedAddr allowed me to use its members directly rather than allocating new strings via method calls on the passed in `net.Addr`.

File: gossip/client_test.go [hunk 34]
```diff
@@ -43,7 +42,7 @@ const (
 func startGossip(t *testing.T) (local, remote *Gossip, stopper *stop.Stopper) {
```
Comment (tschottdorf) [inline gossip/client_test.go]: TODO one line above is obsolete now.

Comment (tschottdorf) [inline gossip/client_test.go]: (will fix in upcoming trivialfix PR)

**Comment by tschottdorf** [comment]
LGTM
